pkg/oam/workload/translate.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workload
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	workloadv1alpha1 "github.com/crossplane/crossplane/apis/workload/v1alpha1"
+)
+
+const (
+	errWrapInKubeApp = "unable to wrap objects in KubernetesApplication"
+)
+
+var (
+	serviceKind       = reflect.TypeOf(corev1.Service{}).Name()
+	serviceAPIVersion = corev1.SchemeGroupVersion.String()
+)
+
+// LabelKey is the label applied to translated workload objects.
+const LabelKey = "workload.oam.crossplane.io"
+
+// KubeAppWrapper wraps a set of translated objects in a KubernetesApplication.
+func KubeAppWrapper(ctx context.Context, w resource.Workload, objs []resource.Object) ([]resource.Object, error) {
+	if objs == nil {
+		return nil, nil
+	}
+
+	app := &workloadv1alpha1.KubernetesApplication{}
+
+	for _, o := range objs {
+		b, err := json.Marshal(o)
+		if err != nil {
+			return nil, errors.Wrap(err, errWrapInKubeApp)
+		}
+
+		kart := workloadv1alpha1.KubernetesApplicationResourceTemplate{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf("%s-%s", o.GetName(), strings.ToLower(o.GetObjectKind().GroupVersionKind().Kind)),
+				Labels: map[string]string{
+					LabelKey: string(w.GetUID()),
+				},
+			},
+			Spec: workloadv1alpha1.KubernetesApplicationResourceSpec{
+				Template: runtime.RawExtension{Raw: b},
+			},
+		}
+
+		app.Spec.ResourceTemplates = append(app.Spec.ResourceTemplates, kart)
+	}
+
+	app.SetName(w.GetName())
+
+	app.Spec.ResourceSelector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			LabelKey: string(w.GetUID()),
+		},
+	}
+
+	return []resource.Object{app}, nil
+}
+
+// ServiceInjector adds a Service object for the first Port on the first
+// Container for the first Deployment observed in a workload translation.
+func ServiceInjector(ctx context.Context, w resource.Workload, objs []resource.Object) ([]resource.Object, error) {
```

Comment:
nit, why is the function name a Noun instead of verb?

Reply:
Mostly because it is passed as middleware in a reconciler. I could be convinced to rename, but I think it reads nicely in this case: https://github.com/crossplane/addon-oam-kubernetes-remote/blob/13c7464dac26cd0dfed88d369f58463fd013d735/pkg/controller/containerizedworkload/containerizedworkload.go#L65